compiler/src/dotty/tools/dotc/transform/init/Objects.scala

```diff
@@ -0,0 +1,1455 @@
+
+      else
+        if ref.klass.isSubClass(receiver.widenSingleton.classSymbol) then
+          report.warning("[Internal error] Unexpected resolution failure: ref.klass = " + ref.klass.show + ", field = " + field.show + Trace.show, Trace.position)
+          Bottom
+        else
+          // This is possible due to incorrect type cast.
+          // See tests/init/pos/Type.scala
+          Bottom
+
+    case fun: Fun =>
+      report.warning("[Internal error] unexpected tree in selecting a function, fun = " + fun.code.show + Trace.show, fun.code)
+      Bottom
+
+    case Bottom =>
+      if field.isStaticObject then ObjectRef(field.moduleClass.asClass)
+      else Bottom
+
+    case RefSet(refs) =>
+      refs.map(ref => select(ref, field, receiver)).join
+  }
+
+  /** Handle assignment `lhs.f = rhs`.
+   *
+   * @param lhs         The value of the object to be mutated.
+   * @param field       The symbol of the target field.
+   * @param rhs         The value to be assigned.
+   * @param rhsTyp      The type of the right-hand side.
+   */
+  def assign(lhs: Value, field: Symbol, rhs: Value, rhsTyp: Type): Contextual[Value] = log("Assign" + field.show + " of " + lhs.show + ", rhs = " + rhs.show, printer, (_: Value).show) {
+    lhs match
+    case fun: Fun =>
+      report.warning("[Internal error] unexpected tree in assignment, fun = " + fun.code.show + Trace.show, Trace.position)
+
+    case Cold =>
+      report.warning("Assigning to cold aliases is forbidden. Calling trace:\n" + Trace.show, Trace.position)
+
+    case Bottom =>
+
+    case RefSet(refs) =>
+      refs.foreach(ref => assign(ref, field, rhs, rhsTyp))
+
+    case ref: Ref =>
+      if ref.hasVar(field) then
+        val addr = ref.varAddr(field)
+        if addr.owner != State.currentObject then
+          errorMutateOtherStaticObject(State.currentObject, addr.owner)
+        else
+          Heap.write(addr, rhs)
+      else
+        report.warning("Mutating a field before its initialization: " + field.show + ". Calling trace:\n" + Trace.show, Trace.position)
+    end match
+
+    Bottom
+  }
+
+  /** Handle new expression `new p.C(args)`.
+   *
+   * @param outer       The value for `p`.
+   * @param klass       The symbol of the class `C`.
+   * @param ctor        The symbol of the target constructor.
+   * @param args        The arguments passsed to the constructor.
+   */
+  def instantiate(outer: Value, klass: ClassSymbol, ctor: Symbol, args: List[ArgInfo]): Contextual[Value] = log("instantiating " + klass.show + ", outer = " + outer + ", args = " + args.map(_.value.show), printer, (_: Value).show) {
+    outer match
+
+    case _ : Fun | _: OfArray  =>
+      report.warning("[Internal error] unexpected outer in instantiating a class, outer = " + outer.show + ", class = " + klass.show + ", " + Trace.show, Trace.position)
+      Bottom
+
+    case value: (Bottom.type | ObjectRef | OfClass | Cold.type) =>
+      // The outer can be a bottom value for top-level classes.
```

Comment:
If `klass` is a top-level class, what would `klass.owner` be?

Reply:
If this is the case, would `klass.owner.isClass` return true on line 817, or it will try the find the enclosing method of the owner? I'm asking this because I'm trying to refine the type for `thisV` in `resolveEnv`.

Reply:
Packages are represented as classes in the compiler --- therefore it will return true for top-level classes. It would be nice to add a comment here.